pkg/input/arm/parser.go

```diff
@@ -0,0 +1,186 @@
+	if strToken, ok := tkn.(stringLiteral); ok {
+		return stringLiteralExpr(strToken), nil
+	}
+
+	// If we reach here, we are building a function expression, because there are
+	// no "direct" identifier dereferences in ARM template expressions. The
+	// identifier is the function name.
+	idToken, ok := tkn.(identifier)
+	if !ok {
+		return nil, newParserError(fmt.Errorf("expected token %#v to be an identifier", tkn))
+	}
+
+	// expect an open paren
+	if p.done() {
+		return nil, newParserError(errors.New("expression cannot terminate with an identifier"))
+	}
+	tkn = p.pop()
+	if _, ok := tkn.(openParen); !ok {
+		return nil, newParserError(fmt.Errorf("expected token %#v to be a paren", tkn))
+	}
+	if p.done() {
+		return nil, newParserError(errors.New("expression cannot terminate with an open paren"))
+	}
+
+	var args []expression
+	for {
+		tkn := p.peek()
+		if _, ok := tkn.(closeParen); ok {
+			p.pop() // pop the close paren
+			expr := functionExpr{name: string(idToken), args: args}
+			if p.done() {
+				return expr, nil
+			}
+			if _, ok := p.peek().(dot); ok {
+				return p.buildPropertyAccessExpression(expr)
+			}
+			return expr, nil
+		}
+
+		// There is a comma between args, so not before the first arg
+		if len(args) > 0 {
+			if _, ok := p.peek().(comma); !ok {
+				return nil, newParserError(fmt.Errorf("expected token %#v to be a comma", tkn))
+			}
+			p.pop() // pop the comma
+		}
+
+		nextArg, err := p.parse()
+		if err != nil {
+			return nil, err
+		}
+		args = append(args, nextArg)
+	}
+}
+
+func (p *parser) buildPropertyAccessExpression(expr expression) (expression, error) {
+	// we only enter this function from parse() if we peeked at a dot, so we know
+	// it gets past here at least once, and so always builds a real property
+	// access expression.
+	if p.done() {
+		return expr, nil
+	}
+	if _, ok := p.peek().(dot); !ok {
+		return expr, nil
+	}
+
+	p.pop() // pop the dot
+	if p.done() {
+		return nil, newParserError(errors.New("expression cannot terminate with a dot"))
+	}
+	tkn := p.pop()
+	nextPropChainElement, ok := tkn.(identifier)
+	if !ok {
+		return nil, newParserError(fmt.Errorf("expected token %#v to be an identifier", tkn))
+	}
+	expr = propertyExpr{obj: expr, property: string(nextPropChainElement)}
+	return p.buildPropertyAccessExpression(expr)
+}
+
+func (p *parser) peek() token {
+	return p.remaining[0]
+}
+
+func (p *parser) pop() token {
+	tkn := p.remaining[0]
+	p.remaining = p.remaining[1:]
+	return tkn
+}
+
+func (p *parser) done() bool {
+	return len(p.remaining) == 0
+}
+
+type expression interface {
+	eval(evalCtx EvaluationContext) (interface{}, error)
+}
+
+type stringLiteralExpr string
+
+func (s stringLiteralExpr) eval(evalCtx EvaluationContext) (interface{}, error) {
+	return string(s), nil
+}
+
+type functionExpr struct {
+	name string
+	args []expression
+}
+
+func (f functionExpr) eval(evalCtx EvaluationContext) (interface{}, error) {
+	impl, ok := evalCtx.funcs[f.name]
```

Comment:
Nitpick: Maybe we should move the `expression` types and these `eval` implementations to a separate `expression.go`?  The types sort of belong here, but having the `eval` in parser seems a bit weird.